### PR TITLE
fix: normalize change_confidence as cosine in [-1, 1], not a 0-145 angle

### LINKED ISSUE #840: change_confidence is normalized as a 0–145 angle, but it's a cosine in [−1, 1]
Hello,

I found what I believe is a units/scale bug in the scoring code.

In` torchange's segment_any_change/anychange.py,` the per-mask score is built as:

```
score = -F.cosine_similarity(t1_mask_embed, t2_mask_embed, dim=0)   # change_confidence
...
keep = change_confidence > angle2cosine(self.change_confidence_threshold)
```

So change_confidence is a negative cosine similarity, range [−1, 1]. 
The change_confidence_threshold=145 hyperparameter is an angle in degrees, which torchange converts with angle2cosine(145) ≈ −0.819 before comparing. 
The stored score is never on a 0–145 scale.

The bug:

`geoai/change_detection.py ` normalizes that value as if it were on the 0–145 angle scale, in two places:

```
_save_probability_mask:

# Normalize confidence: threshold is 145, values above indicate higher confidence
if conf >= 145:                                   # never true (conf  [−1, 1])
    conf_normalized = 0.5 + min(0.5, (conf - 145) / 145)
else:
    conf_normalized = max(0.0, conf / 145 * 0.5)  # negative/145 ,clamped to 0.0
```
    
With weights = {"iou": 0.3, "stability": 0.3, "confidence": 0.35, "area": 0.05}, the 0.35-weighted confidence component contributes 0 for every mask 

```

_save_instance_scores_mask:

change_conf_norm = max(0.0, min(1.0, abs(change_conf) / 200.0))   #  0.004
confidence_score = 0.35 * iou_score + 0.35 * stability_score + 0.3 * change_conf_norm
```

Same issue with a different constant (200) — the 0.3-weighted change term is ~0.004,  negligible.

So, the per-instance "confidence" scores and the probability heatmap measure SAM mask quality only, not how strongly the region changed — regions with very different change magnitudes get near-identical scores.

**Maybe we should Normalize in cosine space:**

```
conf_normalized = max(0.0, min(1.0, (conf + 1.0) / 2.0))   # 0 = identical embeddings, 1 = opposite
(or convert to an angle first, θ = degrees(acos(-conf)), and normalize against change_confidence_threshold, which lives in angle space)
```

### PR DESCRIPTION
**Pull request by giswqs**
## Summary

Fixes #840.

`change_confidence` from torchange's `AnyChange` is a **negative cosine similarity** in `[-1, 1]` (masks are kept when it exceeds `cos(change_confidence_threshold°)`, e.g. `cos(145°) ≈ -0.819`). Three places in `geoai/change_detection.py` normalized it as if it were on a 0–145 (or 0–200) angle scale:

- `_save_probability_mask`: `conf >= 145` was never true, so the 0.35-weighted confidence component was always ~0
- `_save_instance_scores_mask`: `abs(conf) / 200` contributed at most ~0.005 to the 0.3-weighted term
- `_extract_detailed_results` (`combined_confidence`): `conf / 145` clamped to ~0 for the 0.4-weighted term

As a result, per-instance confidence scores and the probability heatmap measured SAM mask quality only — regions with very different change magnitudes got near-identical scores.

## Changes

- Add a shared `_normalize_change_confidence()` helper that maps the raw cosine to `[0, 1]` with the model's decision threshold landing at 0.5 (kept masks map to `(0.5, 1]`, matching the original intent of the probability semantics)
- Read the threshold angle from the torchange model via `_get_change_confidence_threshold()`, so `auto_threshold` (which mutates the model's threshold at runtime) is handled correctly
- Use the helper at all three normalization sites
- Update change confidence report/plot formatting from `.1f` to `.3f` since the raw values are cosines, not large angles
- Add unit tests: threshold maps to 0.5, endpoints map to 0/1, monotonicity, kept masks score above 0.5, input clamping, degenerate 0°/180° thresholds, and threshold read/fallback

## Testing

- `pytest tests/test_change_detection.py` — 35 passed
- `pre-commit run --all-files` — all hooks pass

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved change-confidence normalization so exported masks and detailed results now use consistent, bounded scores.
  * The system now honors the model’s configured confidence threshold automatically, with a safe default when missing.
  * Change-confidence “Mean” values in reports now display with higher precision for more accurate summaries.
* **Tests**
  * Added unit tests covering confidence normalization, threshold reading/fallback behavior, monotonicity, and key edge cases.

### CODE REVIEW

File: geoai/change_detection.py [hunk 1]
```diff
@@ -22,6 +23,38 @@
...
+    Returns:
+        float: Normalized confidence in [0, 1].
+    """
+    conf = float(np.clip(change_confidence, -1.0, 1.0))
+    # Clamp the angle away from 0/180 degrees so both branches below have a
+    # nonzero denominator.
+    threshold_cos = math.cos(math.radians(min(max(threshold_degrees, 1.0), 179.0)))
+    if conf >= threshold_cos:
+        return 0.5 + 0.5 * (conf - threshold_cos) / (1.0 - threshold_cos)
+    return 0.5 * (conf + 1.0) / (threshold_cos + 1.0)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply geoai/change_detection.py:55]: Done in 4b6d4179 — switched to strict `>` to match torchange's keep condition (both branches return exactly 0.5 at the boundary, so behavior is unchanged).

File: tests/test_change_detection.py [hunk 2]
```diff
@@ -452,5 +452,91 @@ def test_report_with_empty_results(self):
...
+    def test_reads_threshold_from_model(self, mock_anychange, mock_download):
+        mock_model = MagicMock()
+        mock_model.change_confidence_threshold = 155
+        mock_anychange.return_value = mock_model
+        mock_download.return_value = "/fake/ckpt.pth"
+
+        from geoai.change_detection import ChangeDetection
+
+        detector = ChangeDetection()
+        self.assertEqual(detector._get_change_confidence_threshold(), 155.0)
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/test_change_detection.py:525]: Done in 4b6d4179 — the threshold is now set on `detector.model` after construction so the test doesn't rely on MagicMock ignoring the init-time `set_hyperparameters` side effect.
